packages/adblocker/src/engine/bucket/cosmetic.ts

```diff
@@ -352,33 +351,74 @@ export default class CosmeticFilterBucket {
     hostname: string;
 
     isFilterExcluded?: (filter: CosmeticFilter) => boolean;
-  }): CosmeticFilter[] {
+  }): {
+    // "rules" is an array of the final result.
+    // If filter has an exception, the filter will not be here.
+    rules: CosmeticFilter[];
+    // "candidates" is the results in the initial stage of matching process before exceptions are applied.
+    // This array contains all filters even there's an exception to help debugging the filtering process in the adblocker library.
+    candidates: CosmeticFilter[];
+    // "exceptions" is the mapping of a filter and an exception.
+    // This data structure allows easy finding of an exception for a filter candidate.
+    // This allows an efficient sequential event emission of a filter and its exception.
+    exceptions: Map<CosmeticFilter, CosmeticFilter>;
+  } {
     // Tokens from `hostname` and `domain` which will be used to lookup filters
     // from the reverse index. The same tokens are re-used for multiple indices.
     const hostnameTokens = createLookupTokens(hostname, domain);
-    const rules: CosmeticFilter[] = [];
+    const candidates: CosmeticFilter[] = [];
     this.htmlIndex.iterMatchingFilters(hostnameTokens, (rule: CosmeticFilter) => {
       if (rule.match(hostname, domain) && !isFilterExcluded?.(rule)) {
-        rules.push(rule);
+        candidates.push(rule);
       }
       return true;
     });
 
+    const exceptions: Map<CosmeticFilter, CosmeticFilter> = new Map();
+
+    if (candidates.length === 0) {
+      return {
+        rules: [],
+        candidates: [],
+        exceptions,
+      };
+    }
+
     // If we found at least one candidate, check if we have unhidden rules.
-    const disabledRules: Set<string> = new Set();
-    if (rules.length !== 0) {
-      this.unhideIndex.iterMatchingFilters(hostnameTokens, (rule: CosmeticFilter) => {
-        if (rule.match(hostname, domain) && !isFilterExcluded?.(rule)) {
-          disabledRules.add(rule.getSelector());
-        }
+    const disabledRules: Map<string, CosmeticFilter> = new Map();
+    this.unhideIndex.iterMatchingFilters(hostnameTokens, (rule: CosmeticFilter) => {
+      if (rule.match(hostname, domain) && !isFilterExcluded?.(rule)) {
+        disabledRules.set(rule.getSelector(), rule);
+      }
 
-        return true;
-      });
+      return true;
+    });
+
+    if (disabledRules.size === 0) {
+      return {
+        rules: candidates,
+        candidates,
+        exceptions,
+      };
     }
 
-    return rules.filter(
-      (rule) => disabledRules.size === 0 || disabledRules.has(rule.getSelector()) === false,
-    );
+    const rules: CosmeticFilter[] = [];
+    let exception: CosmeticFilter | undefined;
```

Comment:
this should be declared in the scope of the for loop

Reply:
Updated.